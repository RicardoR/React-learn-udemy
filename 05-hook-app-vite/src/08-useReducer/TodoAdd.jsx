import {useForm} from "../hooks";

export const TodoAdd = ({onNewTodo}) => {

    const {description, onInputChange, onResetForm} = useForm({
        description: ''
    });

    const createNewTodo = (event) => {
        event.preventDefault();
        if (!description?.trim().length) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            description,
            done: false
        }

        onNewTodo(newTodo);
        onResetForm();
    }

    return (
        <form>
            <input
                type="text"
                placeholder="todo description"
                className="form-control"
                name='description'
                value={description}
                onChange={onInputChange}
            />
            <button
                onClick={createNewTodo}
                className="btn btn-outline-primary mt-1"
                type='submit'
            >
                Add todo
            </button>
        </form>
    );
};

