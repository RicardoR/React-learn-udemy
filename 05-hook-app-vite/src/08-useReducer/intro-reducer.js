const initialState = [
    {
        id: 1,
        todo: 'Comprar pan',
        done: false
    }
];

const todoReducer = (state = initialState, action = {}) => {

    if (action.type === '[TODO] add todo') {
        return [...state, action.payload];
    }


    return state;
}

let todos = todoReducer(initialState);

const newTodo = {
    id: 2,
    todo: 'Comprar vino',
    done: false
};

const addTodoAction = {
    type: '[TODO] add todo',
    payload: newTodo
}

todos = todoReducer(initialState, addTodoAction);

console.log({state: todos});
