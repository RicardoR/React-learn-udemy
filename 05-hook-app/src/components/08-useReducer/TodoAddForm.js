import React from 'react';
import { useForm } from '../../hooks/useForm';
import PropTypes from 'prop-types';

const descriptionField = 'description';

export const TodoAddForm = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, resetInput] = useForm({
    [descriptionField]: '',
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (!description || description.trim().length === 0) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };
    handleAddTodo(newTodo);
    resetInput();
  };

  return (
    <>
      <h4>Agregar Todo</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <div className="input-group">
          <input
            type="text"
            name={descriptionField}
            value={description}
            placeholder="DescripciÃ³n"
            className="form-control"
            autoComplete="off"
            onChange={handleInputChange}
          />
          <div className="input-group-append">
            <button type="onSubmit" className="btn btn-outline-secondary">
              <i className="fa-solid fa-circle-plus"></i>
            </button>
          </div>
        </div>
      </form>
    </>
  );
};

TodoAddForm.propTypes = {
  handleAddTodo: PropTypes.func.isRequired,
};
