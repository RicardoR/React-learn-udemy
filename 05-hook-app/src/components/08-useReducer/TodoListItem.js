import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ({ todo, index, handleComplete, handleRemove }) => {
  return (
    <li
      key={todo.id}
      className="list-group-item d-flex justify-content-between align-items-center"
    >
      <p
        onClick={() => handleComplete(todo.id)}
        className={`cursor-pointer mb-0 
                  ${todo.done ? 'complete' : ''} `}
      >
        {index + 1}. {todo.desc}
      </p>
      <button
        onClick={() => handleRemove(todo.id)}
        className="btn btn-outline-danger"
      >
        <i className="fa fa-trash"></i>
      </button>
    </li>
  );
};

TodoListItem.propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.number.isRequired,
    desc: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired,
  }).isRequired,
  index: PropTypes.number.isRequired,
  handleComplete: PropTypes.func.isRequired,
  handleRemove: PropTypes.func.isRequired,
};
