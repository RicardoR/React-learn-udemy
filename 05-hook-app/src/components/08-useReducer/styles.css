body {
    padding: 50px;
}

.ml-1 {
    margin-left: 1rem;
}

.cursor-pointer {
    cursor: pointer;
}

.complete {
    text-decoration: line-through;
}