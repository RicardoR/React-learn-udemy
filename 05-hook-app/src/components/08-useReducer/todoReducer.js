import {
  ADD_TODO,
  CHANGE_TODO_STATUS,
  REMOVE_TODO,
} from './useReducerConstants';

export const todoReducer = (state = [], action) => {
  switch (action?.type) {
    case ADD_TODO:
      return [...state, action.payload];
    case REMOVE_TODO:
      return state.filter((todo) => todo.id !== action.payload);
    case CHANGE_TODO_STATUS:
      return state.map((todo) => {
        if (todo.id === action.payload) {
          todo.done = !todo.done;
        }
        return todo;
      });
    default:
      return state;
  }
};
