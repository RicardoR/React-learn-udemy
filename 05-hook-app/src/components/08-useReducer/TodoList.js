import React from 'react';
import PropTypes from 'prop-types';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleRemove, handleComplete }) => {
  return (
    <ol className="list-group list-group-flush">
      {todos.map((todo, i) => (
        <TodoListItem
          todo={todo}
          index={i}
          key={todo.id}
          handleComplete={handleComplete}
          handleRemove={handleRemove}
        />
      ))}
    </ol>
  );
};

TodoList.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      desc: PropTypes.string.isRequired,
      done: PropTypes.bool.isRequired,
    }).isRequired
  ),
  handleRemove: PropTypes.func.isRequired,
  handleComplete: PropTypes.func.isRequired,
};
