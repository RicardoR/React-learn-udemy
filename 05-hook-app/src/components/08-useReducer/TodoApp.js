import React, {useReducer } from 'react'
import { todoReducer } from './todoReducer';
import './styles.css';

const initialState = [{
    id: new Date().getTime(),
    desc: 'Learn React',
    done: false
}];

const descriptionField = 'description';

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, initialState);

    const handleSubmit = (e) => { 
        e.preventDefault();

        const newTodo = {
            id: new Date().getTime(),
            desc: e.target[descriptionField].value,
            done: false,
        };

        const action = {
          type: 'ADD_TODO',
          payload: newTodo
        };

        // Esto actualiza el reducer y la app
        dispatch(action);
    }

    const removeTodo = (id) => {
      const action = {
        type: 'REMOVE_TODO',
        payload: id,
      };

      dispatch(action);
    };

    return (
      <div>
        <h1>TodoApp ({todos.length}) </h1>
        <hr />
        <div className="row">
          <div className="col-7">
            <ol className="list-group list-group-flush">
              {todos.map((todo, i) => (
                <li
                  key={todo.id}
                  className="list-group-item d-flex justify-content-between align-items-center"
                >
                  <p className="cursor-pointer mb-0">
                    {i + 1}. {todo.desc}
                  </p>
                  <button className="btn btn-outline-danger">
                    <i className="fa fa-trash"></i>
                  </button>
                </li>
              ))}
            </ol>
          </div>
          <div className="col-5">
            <h4>Agregar Todo</h4>
            <hr />
            <form onSubmit={handleSubmit}>
              <div className="input-group">
                <input
                  type="text"
                  name={descriptionField}
                  placeholder="DescripciÃ³n"
                  className="form-control"
                  autoComplete="off"
                />
                <div className="input-group-append">
                  <button type="onSubmit" className="btn btn-outline-secondary">
                    <i className="fa-solid fa-circle-plus"></i>
                  </button>
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    );
}
