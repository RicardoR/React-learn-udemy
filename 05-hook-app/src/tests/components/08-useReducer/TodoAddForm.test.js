import React from 'react';
import { shallow } from 'enzyme';
import { TodoAddForm } from '../../../components/08-useReducer/TodoAddForm';

describe('TodoAddForm UT', () => {
  const handleAddTodo = jest.fn();
  const wrapper = shallow(<TodoAddForm handleAddTodo={handleAddTodo} />);

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should not call to handleAddTodo wihout values', () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');
    const description = undefined;
    wrapper.find('input').simulate('change', {
      target: { name: 'description', value: description },
    });
    formSubmit({ preventDefault: jest.fn() });
    expect(handleAddTodo).not.toHaveBeenCalled();
  });

  test('should call to handleTodo when there is value in form', () => {
    const preventDefault = jest.fn();
    const description = 'test';
    wrapper.find('input').simulate('change', {
      target: { name: 'description', value: description },
    });

    const formSubmit = wrapper.find('form').prop('onSubmit');

    formSubmit({ preventDefault });
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: description,
      done: false,
    });

    expect(wrapper.find('input').prop('value')).toBe('');
  });
});
