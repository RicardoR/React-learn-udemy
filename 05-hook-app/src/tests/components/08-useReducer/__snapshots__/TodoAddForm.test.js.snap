// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAddForm UT should render correctly 1`] = `
<Fragment>
  <h4>
    Agregar Todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <div
      className="input-group"
    >
      <input
        autoComplete="off"
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="DescripciÃ³n"
        type="text"
        value=""
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-outline-secondary"
          type="onSubmit"
        >
          <i
            className="fa-solid fa-circle-plus"
          />
        </button>
      </div>
    </div>
  </form>
</Fragment>
`;
